src/core/thread/mlr_manager.cpp

```diff
@@ -0,0 +1,362 @@
+        {
+            continue;
+        }
+
+        IgnoreError(mMulticastListenerTable.SetSubscribed(addr->GetAddress(), true));
+    }
+
+    ScheduleSend(0);
+    LogMulticastListenerTable();
+}
+
+void MlrManager::ScheduleSend(uint16_t aDelay)
+{
+    OT_ASSERT(!mCoapPending || mSendDelay == 0);
+
+    VerifyOrExit(!mCoapPending, OT_NOOP);
+
+    if (aDelay == 0)
+    {
+        mSendDelay = 0;
+        SendMulticastListenerRegistration();
+    }
+    else if (mSendDelay == 0 || mSendDelay > aDelay)
+    {
+        mSendDelay = aDelay;
+    }
+
+    ResetTimer();
+exit:
+    return;
+}
+
+void MlrManager::ResetTimer(void)
+{
+    if (mSendDelay == 0 && mReregistrationDelay == 0)
+    {
+        mTimer.Stop();
+    }
+    else if (!mTimer.IsRunning())
+    {
+        mTimer.Start(kTimerInterval);
+    }
+}
+
+void MlrManager::SendMulticastListenerRegistration()
+{
+    otError          error   = OT_ERROR_NONE;
+    Mle::MleRouter & mle     = Get<Mle::MleRouter>();
+    Coap::Message *  message = nullptr;
+    Ip6::MessageInfo messageInfo;
+    IPv6AddressesTlv addressesTlv;
+    uint8_t          num;
+
+    VerifyOrExit(!mCoapPending, error = OT_ERROR_BUSY);
+    VerifyOrExit(mle.IsAttached(), error = OT_ERROR_INVALID_STATE);
+    VerifyOrExit(mle.IsFullThreadDevice() || mle.GetParent().IsThreadVersion1p1(), error = OT_ERROR_INVALID_STATE);
+    VerifyOrExit(Get<BackboneRouter::Leader>().HasPrimary(), error = OT_ERROR_INVALID_STATE);
+
+    num = mMulticastListenerTable.Count(MulticastListenerTable::kFilterDirty);
+    VerifyOrExit(num > 0, error = OT_ERROR_NOT_FOUND);
+    if (num > kIPv6AddressesNumMax)
+    {
+        num = kIPv6AddressesNumMax;
+    }
+
+    VerifyOrExit((message = Get<Coap::Coap>().NewMessage()) != nullptr, error = OT_ERROR_NO_BUFS);
+
+    message->Init(OT_COAP_TYPE_CONFIRMABLE, OT_COAP_CODE_POST);
+    SuccessOrExit(message->SetToken(Coap::Message::kDefaultTokenLength));
+    SuccessOrExit(message->AppendUriPathOptions(OT_URI_PATH_MLR));
+    SuccessOrExit(message->SetPayloadMarker());
+
+    addressesTlv.Init();
+    addressesTlv.SetLength(sizeof(Ip6::Address) * num);
+    SuccessOrExit(error = message->Append(&addressesTlv, sizeof(addressesTlv)));
+
+    {
+        MulticastListenerTable::Iterator iter(GetInstance(), MulticastListenerTable::kFilterDirty);
+        for (uint8_t i = 0; i < num; i++, iter.Advance())
+        {
+            OT_ASSERT(!iter.IsDone());
+            SuccessOrExit(error = message->Append(&iter.GetMulticastListener().GetAddress(), sizeof(Ip6::Address)));
+        }
+    }
+
+    if (!mle.IsFullThreadDevice() && mle.GetParent().IsThreadVersion1p1())
+    {
+        uint8_t pbbrServiceId;
+
+        SuccessOrExit(error = Get<BackboneRouter::Leader>().GetServiceId(pbbrServiceId));
+        SuccessOrExit(error = mle.GetServiceAloc(pbbrServiceId, messageInfo.GetPeerAddr()));
+    }
+    else
+    {
+        messageInfo.GetPeerAddr().SetToAnycastLocator(mle.GetMeshLocalPrefix(),
```

Comment:
nit: update `SetToAnycastLocator` to be `SetToRoutingLocator` as the address constructed is RLOC, not ALOC.